inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/reorder_weights_opt.cl

```diff
@@ -0,0 +1,177 @@
+#if OFM_FIRST
+#if OUTPUT_GROUPED
+    const int i = (g_io % (OUTPUT_IFM_NUM / SECOND_BLOCK_SIZE)) * SECOND_BLOCK_SIZE;
+    const int g = (g_io / (OUTPUT_IFM_NUM / SECOND_BLOCK_SIZE));
+#else
+    const int g = 0;
+    const int i = g_io * SECOND_BLOCK_SIZE;
+#endif  // OUTPUT_GROUPED
+    const int o_blocked = (int)get_group_id(2) * FIRST_BLOCK_SIZE;
+    const int o = o_blocked + lid;
+    const int i_blocked = i;
+#else  // OFM_FIRST
+#if OUTPUT_GROUPED
+    const int o = (g_io % (OUTPUT_OFM_NUM / SECOND_BLOCK_SIZE)) * SECOND_BLOCK_SIZE;
+    const int g = (g_io / (OUTPUT_OFM_NUM / SECOND_BLOCK_SIZE));
+#else
+    const int g = 0;
+    const int o = g_io * SECOND_BLOCK_SIZE;
+#endif  // OUTPUT_GROUPED
+    const int i_blocked = (int)get_group_id(2) * FIRST_BLOCK_SIZE;
+    const int i = i_blocked + lid;
+    const int o_blocked = o;
+#endif  // OFM_FIRST
+
+    const int zyx = get_global_id(1);
+    const int x = zyx % OUTPUT_SIZE_X;
+#if (OUTPUT_DIMS - OUTPUT_GROUPED) == 5
+    const int y = zyx / OUTPUT_SIZE_X % OUTPUT_SIZE_Y;
+    const int z = zyx / OUTPUT_SIZE_X / OUTPUT_SIZE_Y;
+#else
+    const int y = zyx / OUTPUT_SIZE_X;
+    const int z = 0;
+#endif  // (OUTPUT_DIMS - OUTPUT_GROUPED) == 5
+
+    int input_idx = GET_INDEX(INPUT0_GET_INDEX, IDX_ORDER);
+    const int output_idx = GET_INDEX(OUTPUT_GET_INDEX, BLOCK_IDX_ORDER);
+
+// if (lid == 0)
+//     printf("Input idx: %d %d %d %d %d [%d]\nOutput idx: %d %d %d %d %d [%d]\n", IDX_ORDER, input_idx, BLOCK_IDX_ORDER, output_idx);
+
+// #if INPUT0_SIZE_Y == 7 && INPUT0_SIZE_X == 7
+//     if (lid == 0)
+//         printf("%d %d %d %d: %d\n", o, i, y, x, output_idx);
+// #endif
+
+#if SECOND_BLOCK_SIZE == 1
+    const OUTPUT_TYPE val = TO_OUTPUT_TYPE(input[input_idx]);
+#else
+    OUTPUT_VEC_TYPE val = 0;
+    __attribute__((opencl_unroll_hint))
+    for (int i = 0; i < SECOND_BLOCK_SIZE; i++) {
+        val[i] = TO_OUTPUT_TYPE(input[input_idx]);
+        input_idx += PITCH;
+    }
+#endif  // SECOND_BLOCK_SIZE
+#if OUTPUT_LEFTOVERS
+#if OFM_FIRST
+    const bool doWrite = o < OUTPUT_OFM_NUM;
+    if (o_blocked >= OUTPUT_OFM_NUM - FIRST_BLOCK_SIZE) {
+#else
+    const bool doWrite = i < OUTPUT_IFM_NUM;
+    if (i_blocked >= OUTPUT_IFM_NUM - FIRST_BLOCK_SIZE) {
+#endif  // OFM_FIRST
+#if SECOND_BLOCK_SIZE > 1
+        __attribute__((opencl_unroll_hint))
+        for (int b = 0; b < SECOND_BLOCK_SIZE; b++)
+            if (doWrite)
+                output[output_idx + b * SECOND_SIZE + lid] = val[b];
+#else
+            if (doWrite)
+                output[output_idx + lid] = val;
+#endif  // SECOND_BLOCK_SIZE > 1
+    }
+    else
+#endif  // OUTPUT_LEFTOVERS
+    {
+#if SECOND_BLOCK_SIZE == 8
+        OUTPUT_BLOCK_WRITE8(output, output_idx, val);
+#elif SECOND_BLOCK_SIZE == 4
+        OUTPUT_BLOCK_WRITE4(output, output_idx, val);
+#elif SECOND_BLOCK_SIZE == 2
+        OUTPUT_BLOCK_WRITE2(output, output_idx, val);
+#else
+        OUTPUT_BLOCK_WRITE(output, output_idx, val);
+#endif
```

Comment:
Maybe it makes sence to do one OUTPUT_BLOCK_WRITE macro using CAT and SECOND_BLOCK_SIZE?

Reply:
Done